[Profiler] Defer KinetoEvent and GenericTraceActivity creation to post processing.

robieta opened the pull request:
Summary:
This is the first of the optimizing changes. One of the issues with kineto sometimes being unavailable is we cannot use it as a storage mechanism. KinetoEvent currently fills this role, however KinetoEvent is VERY expensive. A second issue is that because we currently write to two objects, we hold the state lock for the duration of both event creations which is not ideal.

This applies the following optimizations:
1) Intermediate data is stored in a deque in KinetoThreadLocalState, which saves a data->KinetoObserverContext->KinetoEvent double copy. The new KinetoObserverContext just holds a pointer to the element in the deque.
2) OpEventData is much lighter weight (though still far from ideal)

Test Plan:
Script: P470970719
Result: P470970794
For the base case (no special flags), 40% reduction in the `profiler_kineto` portion of the overhead.

Differential Revision: D32691800

aaronenyeshi reviewed (approved):
LGTM, stamping! Please check if any CI failures are caused by this.

Commit:
[Profiler] Defer KinetoEvent and GenericTraceActivity creation to post processing. (#71539)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/71539

This is the first of the optimizing changes. One of the issues with kineto sometimes being unavailable is we cannot use it as a storage mechanism. KinetoEvent currently fills this role, however KinetoEvent is VERY expensive. A second issue is that because we currently write to two objects, we hold the state lock for the duration of both event creations which is not ideal.

This applies the following optimizations:
1) Intermediate data is stored in a deque in KinetoThreadLocalState, which saves a data->KinetoObserverContext->KinetoEvent double copy. The new KinetoObserverContext just holds a pointer to the element in the deque.
2) OpEventData is much lighter weight (though still far from ideal)

Test Plan:
Script: P470970719
Result: P470970794
For the base case (no special flags), 40% reduction in the `profiler_kineto` portion of the overhead.

Reviewed By: aaronenyeshi

Differential Revision: D32691800

fbshipit-source-id: d38d551d94689ac3da68f35d25e4dc1767c7a1d3

anjali411 commented:
@robieta could you please add the release notes labels?

robieta commented:
> @robieta could you please add the release notes labels?

I added `release notes: profiler`. Is there more I need to do?

anjali411 commented:
> > @robieta could you please add the release notes labels?
> 
> I added `release notes: profiler`. Is there more I need to do?

ahh yeah you can optionally add a `topic: ...` label (performance/bug-fix/improvement/new feature/...)
